From 8ced2a0bdcda5731657bac39f399fe270defc803 Mon Sep 17 00:00:00 2001
From: "Jose M. Guisado Gomez" <guigom@riseup.net>
Date: Wed, 24 Jun 2020 15:57:07 +0200
Subject: [PATCH] nftables plugin: add libnftnl dependency check

Removes verbatim library includes for libnftnl.

This is a first approach about checking for libnftnl, the dependency
for the nftables plugin to work and communicate via netlink messages to
the kernel and getting the counters.

This is an adaptation of --with-iptables configure.ac section. In
summary it will:
    * ask if libnftnl exists to pkg-config
    * check if it's running on a Linux machine
    * check for some necessary headers and symbols from libnftnl used in
      this plugin
---
 Makefile.am  |  4 ++--
 configure.ac | 58 +++++++++++++++++++++++++++++++++++++++++++++++++++-
 2 files changed, 59 insertions(+), 3 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 5f1eb81f22..82583b4fc5 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1543,9 +1543,9 @@ endif
 if BUILD_PLUGIN_NFTABLES
 pkglib_LTLIBRARIES += nftables.la
 nftables_la_SOURCES = src/nftables.c
-nftables_la_CPPFLAGS = $(AM_CPPFLAGS) -I/usr/local/include
+nftables_la_CPPFLAGS = $(AM_CPPFLAGS) $(BUILD_WITH_LIBNFTNL_CFLAGS)
 nftables_la_LDFLAGS = $(PLUGIN_LDFLAGS)
-nftables_la_LIBADD = -lnftnl
+nftables_la_LIBADD = $(BUILD_WITH_LIBNFTNL_LIBS)
 endif
 
 if BUILD_PLUGIN_NGINX
diff --git a/configure.ac b/configure.ac
index bee9c7fee0..bb494592b9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -3981,6 +3981,62 @@ AC_SUBST([BUILD_WITH_LIBNETSNMPAGENT_LDFLAGS])
 AC_SUBST([BUILD_WITH_LIBNETSNMPAGENT_LIBS])
 # }}}
 
+# --with-libnftnl {{{
+AC_ARG_WITH([libnftnl],
+  [AS_HELP_STRING([--with-libnftnl@<:@=PREFIX@:>@], [Path to libnftnl. ])],
+  [
+    if test "x$withval" = "xyes"; then
+      with_libnftnl="yes"
+     else if test "x$withval" = "xno"; then
+       with_libnftnl="no"
+     else
+       if test -d "$withval/include"; then
+         with_libnftnl_cflags="-I$withval/include"
+         with_libnftnl_libs="-L$withval/lib -lnftnl"
+         with_libnftnl="yes"
+       else
+         AC_MSG_ERROR("no such directory: $withval/include")
+       fi
+     fi; fi
+  ],
+  [
+    if test "x$ac_system" = "xLinux"; then
+      with_libnftnl="yes"
+    else
+      with_libnftnl="no (Linux only library)"
+    fi
+  ]
+)
+
+if test "x$with_libnftnl" = "xyes"; then
+  if $PKG_CONFIG --exists libnftnl 2>/dev/null; then
+    with_libnftnl_cflags="$with_libnftnl_ldflags `$PKG_CONFIG --cflags libnftnl`"
+    with_libnftnl_libs="$with_libnftnl_libs `$PKG_CONFIG --libs libnftnl`"
+  fi
+
+  AC_CHECK_HEADERS([libnftnl/expr.h libnftnl/rule.h libnftnl/udata.h],
+    [],
+    [
+      with_libnftnl="no"
+      break
+    ])
+
+  AC_CHECK_LIB([nftnl], [nftnl_nlmsg_build_hdr],
+    [with_libnftnl="yes"],
+    [with_libnftnl="no (symbol 'nftnl_nlmsg_build_hdr' not found)"],
+    [$with_libnftnl_libs]
+  )
+fi
+
+if test "x$with_libnftnl" = "xyes"; then
+  BUILD_WITH_LIBNFTNL_CFLAGS="$with_libnftnl_cflags"
+  BUILD_WITH_LIBNFTNL_LIBS="$with_libnftnl_libs"
+fi
+AC_SUBST([BUILD_WITH_LIBNFTNL_CFLAGS])
+AC_SUBST([BUILD_WITH_LIBNFTNL_LIBS])
+AM_CONDITIONAL([HAVE_LIBNFTNL], [test "x$with_libnftnl" = "xyes"])
+# }}}
+
 # --with-liboping {{{
 AC_ARG_WITH([liboping],
   [AS_HELP_STRING([--with-liboping@<:@=PREFIX@:>@], [Path to liboping.])],
@@ -7095,7 +7151,7 @@ AC_PLUGIN([netapp],              [$with_libnetapp],           [NetApp plugin])
 AC_PLUGIN([netlink],             [$with_libmnl],              [Enhanced Linux network statistics])
 AC_PLUGIN([network],             [yes],                       [Network communication plugin])
 AC_PLUGIN([nfs],                 [$plugin_nfs],               [NFS statistics])
-AC_PLUGIN([nftables],            [yes],                       [nftables rules statistics])
+AC_PLUGIN([nftables],            [$with_libnftnl],            [nftables rules statistics])
 AC_PLUGIN([nginx],               [$with_libcurl],             [nginx statistics])
 AC_PLUGIN([notify_desktop],      [$with_libnotify],           [Desktop notifications])
 AC_PLUGIN([notify_email],        [$with_libesmtp],            [Email notifier])
